pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -670,16 +672,34 @@ impl LayerInner {
                 Ok(ResidentOrWantedEvicted::Resident(res))
             };
 
-            let locked = self.inner.get_or_init(download).await?;
+            let (weak, _permit) = {
+                // should we be able to give the permit to the `get_or_init`? would make sense.
+                drop(permit.take());
+                let mut locked = self.inner.get_or_init(download).await?;
 
-            if let Some(strong) = Self::get_or_apply_evictedness(Some(locked), &self.wanted_evicted)
-            {
-                return Ok(strong);
-            }
+                if let Some(strong) = locked.get() {
+                    return Ok(strong);
+                } else {
+                    // path to here:
+                    // 1. the on_downloaded_layer_drop is stuck on spawn_blocking queue
+                    // 2. is there something else?
+                    //
+                    // reset the contents, deactivating the eviction and causing a
+                    // EvictionCancelled::LostToDownload.
+                    locked.take_and_deinit()
```

Comment:
Responding to your earlier comment here https://github.com/neondatabase/neon/pull/5742#pullrequestreview-1706624340:

> by making downloads amend existing layer state to cancel eviction

I did not implement this. It could be implemented easily by having the `locked.get()` do instead of just `clone()` or `upgrade()` write any `ResidentOrWantedEvicted::WantedEvicted => ResidentOrWantedEvicted::Resident`. I did not consider this initially because I did want even the wrong evictions to complete.

But this would be easy to implement. Will do that.

Reply:
66a58ca58

Reply:
I think the atomic `wanted_evicted` is still good for avoiding an unconditional spawn_blocking even if it always mirrors the `ResidentOrWantedEvicted`.